sophia/src/serializer.rs

```diff
@@ -1,217 +1,355 @@
 ///
-/// See also [`TripleSink`].
-///
-/// [`TripleSink`]: ../triple/stream/trait.TripleSink.html
-///
-pub trait TripleWriter<W: io::Write>: TripleSink<Outcome = ()> + Sized {
-    type Config;
+/// [`TripleSource`]: ../triple/stream/trait.TripleSource.html
+/// [`Graph`]: ../graph/trait.Graph.html
+pub trait TripleSerializer<T> {
+    type Sink: TripleSink;
 
-    fn new(write: W, config: Self::Config) -> Self;
+    /// A sink serializing into `target` the triples fed to it.
+    fn sink(&self, target: T) -> Self::Sink;
 
-    /// Serialize the triples from the given source.
-    fn write<TS, T>(&mut self, mut source: TS) -> Result<(), StreamError<TS::Error, Self::Error>>
+    #[inline]
+    /// Convenient shortcut method for serializing a [`TripleSource`].
+    ///
+    /// [`TripleSource`]: ../triple/stream/trait.TripleSource.html
+    fn serialize_triples_in<TS>(&self, mut source: TS, target: T) -> TSResult<Self, T, TS::Error>
     where
         TS: TripleSource,
     {
-        source.in_sink(self)
+        source.in_sink(&mut self.sink(target))
     }
 
-    /// Serialize the given graph.
-    fn write_graph<G>(&mut self, graph: &mut G) -> Result<(), StreamError<G::Error, Self::Error>>
+    /// Serialize a whole [`Graph`] into `target`.
+    ///
+    /// While this method has a default implementation relying on [`serialize_triples_in`],
+    /// some implementation may override it in order to better use the structure of the Graph.
+    ///
+    /// [`Graph`]: ../graph/trait.Graph.html
+    /// [`serialize_triples_in`]: #method.serialize_triples_in
+    #[inline]
+    fn serialize_graph_in<G>(&self, graph: &G, target: T) -> TSResult<Self, T, G::Error>
     where
         G: Graph,
     {
-        graph.triples().in_sink(self)
-    }
-
-    /// Serialize the given triple.
-    fn write_triple<T>(&mut self, t: &T) -> Result<(), StreamError<Infallible, Self::Error>>
-    where
-        T: Triple,
-    {
-        let mut source = vec![[t.s(), t.p(), t.o()]].into_iter().as_triple_source();
-        source.in_sink(self)
+        self.serialize_triples_in(graph.triples(), target)
     }
 }
 
-/// An extension of the `TripleSink` trait,
-/// dedicated to serialization to strings,
-/// with methods more explicitly named.
-///
-/// See also [`TripleSink`].
-///
-/// [`TripleSink`]: ../triple/stream/trait.TripleSink.html
+/// A triple stringifier serializes triples into a string.
+/// The triples may come from a [`TripleSource`] or a [`Graph`].
 ///
-pub trait TripleStringifier: TripleSink<Outcome = String> + Sized {
-    type Config;
-
-    fn new(config: Self::Config) -> Self;
-
-    /// Stringify the triples from the given source.
-    fn stringify<TS, T>(
-        &mut self,
-        mut source: TS,
-    ) -> Result<String, StreamError<TS::Error, Self::Error>>
+/// [`TripleSource`]: ../triple/stream/trait.TripleSource.html
+/// [`Graph`]: ../graph/trait.Graph.html
+pub trait TripleStringifier {
```

Comment:
Regarding my proposal at line 21 this would be:

```rust
fn vec_to_string(v: Vec<u8>) -> String {
    String::from_utf8_lossy(&v).to_owned()
}

trait TripleStringifier: TripleSerializer<Outcome=Vec<u8>> {
    fn stringify_term<TD: TermData>(self, t: &Term<TD>) -> Result<String, Self::Error> {
        let mut ser = self;
        ser.serialize_term(t)?;
        ser.finish().map(vec_to_string)
    }

    ...
}
```

Maybe add some shorthand functions for `Self: Default` ...

Reply:
I'm not a big fan either. Here are alternative solutions I considered, by decreasing order of preference:

* have `TripleStringifier` (resp. `QuadStringifier`) automatically implemented by any impl of `TripleSerializer` (resp. `QuadSerializer`) supporting `&mut Vec<u8>` (which includes those supporting any `Writer`). I tried  to do it that way, but I failed. The problem is that you need to declare a lifetime for the mut reference to `Vec<u8>`, and this proves to be tricky. I even tried something with a HRTB (`for <'a> ...`) but that failed as well.

* have a proper `derive` macro, so that one could simply write `[derive(TripleStringifier)]` above their type implementing `TripleSerializer`. But I don't know procedural macros.

* have hacky pseudo-derive using a standard macro. This is where we are.

But if you manage to get one the solutions above, I'm all for it.

Reply:
This clearly depends on the `Serializer` trait. Of course I'll spent time on that to get a working solution. But lets settle this for now and focus on the main trait.